pkg/nsx/services/nsxserviceaccount/store.go

```diff
@@ -0,0 +1,130 @@
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	nsxutil "github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+func queryPrincipalIdentity(service *NSXServiceAccountService, wg *sync.WaitGroup, fatalErrors chan error) {
+	defer wg.Done()
+	queryParam := common.QueryTagCondition(common.ResourceTypePrincipalIdentity, service.NSXConfig.CoeConfig.Cluster)
+	var cursor *string = nil
+	pageSize := common.PageSize
+	for {
+		response, err := service.NSXClient.MPQueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
+		err = common.TransError(err)
+		if _, ok := err.(nsxutil.PageMaxError); ok == true {
+			common.DecrementPageSize(&pageSize)
+			continue
+		}
+		if err != nil {
+			fatalErrors <- err
+		}
+		for _, g := range response.Results {
+			a, err := Converter.ConvertToGolang(g, mpmodel.PrincipalIdentityBindingType())
+			if err != nil {
+				for _, e := range err {
+					fatalErrors <- e
+				}
+			}
+			c, _ := a.(mpmodel.PrincipalIdentity)
+			err2 := service.PrincipalIdentityStore.Add(c)
+			if err2 != nil {
+				fatalErrors <- err2
+			}
+		}
+		cursor = response.Cursor
+		if cursor == nil {
+			break
+		}
+		c, _ := strconv.Atoi(*cursor)
+		if int64(c) >= *response.ResultCount {
+			break
+		}
+	}
+}
+
+func queryClusterControlPlane(service *NSXServiceAccountService, wg *sync.WaitGroup, fatalErrors chan error) {
+	defer wg.Done()
+	queryParam := common.QueryTagCondition(common.ResourceTypeClusterControlPlane, service.NSXConfig.CoeConfig.Cluster)
+	var cursor *string = nil
+	pageSize := common.PageSize
+	for {
+		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
+		err = common.TransError(err)
+		if _, ok := err.(nsxutil.PageMaxError); ok == true {
+			common.DecrementPageSize(&pageSize)
+			continue
+		}
+		if err != nil {
+			fatalErrors <- err
+		}
+		for _, g := range response.Results {
+			a, err := Converter.ConvertToGolang(g, model.ClusterControlPlaneBindingType())
+			if err != nil {
+				for _, e := range err {
+					fatalErrors <- e
+				}
+			}
+			c, _ := a.(mpmodel.PrincipalIdentity)
+			err2 := service.PrincipalIdentityStore.Add(c)
+			if err2 != nil {
+				fatalErrors <- err2
+			}
+		}
+		cursor = response.Cursor
+		if cursor == nil {
+			break
+		}
+		c, _ := strconv.Atoi(*cursor)
+		if int64(c) >= *response.ResultCount {
+			break
+		}
+	}
+}
+
+// KeyFunc returns the key of the object.
+func KeyFunc(obj interface{}) (string, error) {
+	switch v := obj.(type) {
+	case model.ClusterControlPlane:
+		return *v.Id, nil
+	case mpmodel.PrincipalIdentity:
+		return *v.Name, nil
+	default:
+		return "", errors.New("keyFunc doesn't support unknown type")
+	}
+}
+
+// indexFunc is used to get index of a resource, usually, which is the UID of the CR controller reconciles,
+// index is used to filter out resources which are related to the CR
+func IndexFunc(obj interface{}) ([]string, error) {
```

Comment:
Keep it lowercase since it is only used inside the package?

Reply:
Fixed.